src/pg/replication.cr

```diff
@@ -0,0 +1,353 @@
+      @format = Format.new(sized.read_bytes(Int8, IO::ByteFormat::NetworkEndian))
+      column_count = sized.read_bytes(Int16, IO::ByteFormat::NetworkEndian)
+      @column_formats = Array.new(column_count) do
+        sized.read_bytes(Int16, IO::ByteFormat::NetworkEndian)
+      end
+      sized.close
+    end
+
+    enum Format : Int8
+      Text   = 0
+      Binary = 1
+    end
+  end
+
+  struct CopyData < Frame
+    getter data : XLogData | KeepAlive
+
+    def initialize(io : IO)
+      size = io.read_bytes(Int32, IO::ByteFormat::NetworkEndian)
+      case byte = io.read_byte
+      when Nil
+        raise IO::EOFError.new("Connection was unexpectedly terminated")
+      when 'w'
+        @data = XLogData.new(io)
+      when 'k'
+        @data = KeepAlive.new(io)
+      else
+        raise Error.new("Unexpected CopyData byte marker: 0x#{byte.to_s(16)} (#{byte.chr.inspect})")
+      end
+    end
+  end
+
+  struct XLogData
+    getter start : Int64
+    getter end : Int64
+    getter timestamp : Time
+    getter message : WALMessage
+
+    def initialize(io : IO)
+      @start = io.read_bytes(Int64, IO::ByteFormat::NetworkEndian)
+      @end = io.read_bytes(Int64, IO::ByteFormat::NetworkEndian)
+      @timestamp = TimeParser.call(io)
+      @message = WALMessage.new(io)
+    end
+  end
+
+  struct KeepAlive
+    getter end : Int64
+    getter timestamp : Time
+    getter? response_expected : Bool
+
+    def initialize(io : IO)
+      @end = io.read_bytes(Int64, IO::ByteFormat::NetworkEndian)
+      @timestamp = TimeParser.call(io)
+      @response_expected = io.read_bytes(Int8, IO::ByteFormat::NetworkEndian) == 1
+    end
+  end
+
+  abstract struct WALMessage
+    def self.new(io : IO) : self
+      case byte = io.read_byte
+      when Nil
+        raise IO::EOFError.new("Connection was unexpectedly terminated")
+      when 'B'
+        Begin.new(io)
+      when 'C'
+        Commit.new(io)
+      when 'R'
+        Relation.new(io)
+      when 'I'
+        Insert.new(io)
+      when 'U'
+        Update.new(io)
+      else
+        raise Error.new("Unexpected WAL message byte marker: 0x#{byte.to_s(16)} (#{byte.chr.inspect})")
+      end
+    end
+
+    protected def read(io : IO, int : Int.class)
+      io.read_bytes int, IO::ByteFormat::NetworkEndian
+    end
+
+    protected def read_string(io : IO) : String
+      io.read_line('\0', chomp: true)
+    end
+
+    protected def read_bytes(io : IO) : Bytes
+      bytes = Bytes.new(read(io, Int32))
+      io.read_fully bytes
+      bytes
+    end
+
+    protected def read_time(io : IO) : Time
+      TimeParser.call(io)
+    end
+
+    protected def read_tuple_data(io) : TupleData
+      column_count = read(io, Int16)
+      Array.new(column_count) do
+        case byte = io.read_byte
+        when Nil
+          raise IO::EOFError.new("Connection was unexpectedly terminated")
+        when 'n'
+          nil
+        when 'u'
+          UnchangedTOASTValue.new
```

Comment:
I haven't tested this yet, but the docs say that [an unchanged TOASTed value is not sent](https://www.postgresql.org/docs/current/protocol-logicalrep-message-formats.html#PROTOCOL-LOGICALREP-MESSAGE-FORMATS-TUPLEDATA), so I'm assuming it's just a marker to indicate that the previous value is unchanged to avoid sending huge strings over the wire unnecessarily.